### PR TITLE
Add caching for k8s build and shard tests

### PR DESCRIPTION
**Pull request by as-com**
This brings down (cached) CI runs from ~35 minutes to ~22 minutes, while also running more e2e tests.

cc @trozet @dcbw

### CODE REVIEW

**Comment by trozet** [comment]
looks like the total number of cases ran for this patch are 44. Previously I see we were running 38. Is that what you expect @as-com ?

**Comment by as-com** [comment]
@trozet yes, this was intentional - this PR causes more test cases to run. Just for the record, here are the test cases that I have seen fail either every time or intermittently:

## [sig-network] [Feature:PerformanceDNS][Serial]
(fails with etcd running out of space, causing all tests afterwards to fail)

## [sig-network] Networking IPerf IPv4 [Experimental] [Feature:Networking-IPv4] [Slow] [Feature:Networking-Performance] 
[It] should transfer ~ 1GB onto the service endpoint 1 servers (maximum of 1 clients)
## [sig-network] Networking IPerf IPv6 [Experimental] [Feature:Networking-IPv6] [Slow] [Feature:Networking-Performance] [LinuxOnly] 
[It] should transfer ~ 1GB onto the service endpoint 1 servers (maximum of 1 clients)  
## [sig-network] [Feature:IPv6DualStackAlphaFeature] [LinuxOnly] 
[It] should be able to reach pod on ipv4 and ipv6 ip [Feature:IPv6DualStackAlphaFeature:Phase2]
[It] should have ipv4 and ipv6 node podCIDRs 
[It] should have ipv4 and ipv6 internal node ip 
[It] should create pod, add ipv6 and ipv4 ip to pod ips 
## [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client 
[It] should allow ingress access from namespace on one named port [Feature:NetworkPolicy] 
[It] should allow egress access on one named port [Feature:NetworkPolicy] 
[It] should allow ingress access from updated pod [Feature:NetworkPolicy] 
[It] should allow ingress access on one named port [Feature:NetworkPolicy]  
[It] should enforce policy to allow traffic only from a pod in a different namespace based on PodSelector and NamespaceSelector [Feature:NetworkPolicy] 
[It] should allow egress access to server in CIDR block [Feature:NetworkPolicy]
[It] should enforce policy to allow traffic from pods within server namespace based on PodSelector [Feature:NetworkPolicy]
[It] should enforce multiple ingress policies with ingress allow-all policy taking precedence [Feature:NetworkPolicy]
[It] should enforce policy to allow traffic only from a different namespace, based on NamespaceSelector [Feature:NetworkPolicy]
## [sig-network] NoSNAT [Feature:NoSNAT] [Slow]
[It] Should be able to send traffic between Pods without SNAT 
## [sig-network] Services 
[It] should allow pods to hairpin back to themselves through services
[It] should handle load balancer cleanup finalizer for service [Slow] 
[It] should have session affinity work for LoadBalancer service with ESIPP off [Slow] [DisabledForLargeClusters] 
[It] should be able to switch session affinity for service with type clusterIP 
[It] should be able to switch session affinity for NodePort service 
[It] should have session affinity work for service with type clusterIP 
[It] should be rejected when no endpoints exist 
[It] should be able to switch session affinity for LoadBalancer service with ESIPP on [Slow] [DisabledForLargeClusters] 
[It] should have session affinity work for NodePort service 
[It] should have session affinity work for LoadBalancer service with ESIPP on [Slow] [DisabledForLargeClusters] 
[It] should be able to switch session affinity for LoadBalancer service with ESIPP off [Slow] [DisabledForLargeClusters] 
## [sig-network] Networking 
[It] should provide Internet connection for containers [Feature:Networking-IPv6][Experimental][LinuxOnly] 
## [sig-network] ClusterDns [Feature:Example] 
[It] should create pod that uses dns 
## [sig-network] DNS configMap federations [Feature:Federation] 
[It] should be able to change federation configuration [Slow][Serial] 
## [sig-network] DNS configMap nameserver [Feature:Networking-IPv6] [LinuxOnly] 
Change stubDomain [It] should be able to change stubDomain configuration [Slow][Serial] 
Forward PTR lookup [It] should forward PTR records lookup to upstream nameserver [Slow][Serial] 
Forward external name lookup [It] should forward externalname lookup to upstream nameserver [Slow][Serial] 
## [sig-network] DNS configMap nameserver [IPv4] 
Forward external name lookup [It] should forward externalname lookup to upstream nameserver [Slow][Serial] 
Forward PTR lookup [It] should forward PTR records lookup to upstream nameserver [Slow][Serial]

**Comment by danwinship** [comment]
> [sig-network] Networking IPerf IPv4 [Experimental] [Feature:Networking-IPv4] [Slow] [Feature:Networking-Performance]

Given our resources constraints probably we shouldn't run any tests tagged `[Slow]`?

Also, I suspect `[Experimental]` may mean that it's not run upstream and may not be stable. Probably best to not run any tests with that tag.

> [sig-network] [Feature:IPv6DualStackAlphaFeature] [LinuxOnly]

We don't support dual-stack at this time, only single-stack IPv6.



> [It] should allow pods to hairpin back to themselves through services

Service hairpinning is known to not work yet, #817 

>  [It] should be able to switch session affinity for NodePort service

Likewise session affinity, #819 

>  [It] should be rejected when no endpoints exist

Tim is working on that one, #928 

> [sig-network] ClusterDns [Feature:Example]

This may assume particular things about the cluster's DNS setup that may not be true in our test cluster.

> [sig-network] DNS configMap federations [Feature:Federation]

We do not have a federated multi-cluster. You should not run this test.

> [sig-network] DNS configMap nameserver ...

I'm not sure what these tests do, but again it's possible they won't work in our environment

**Comment by danwinship** [comment]
> We don't support dual-stack at this time, only single-stack IPv6.

and actually, the kind environment we set up for tests only supports IPv4. We'd need to set up a separate test cluster to run IPv6 tests

**Comment by trozet** [comment]
@girishmg PTAL

**Comment by dcbw** [comment]
@trozet @as-com can you rebase? Probably conflicts with the small PR to keep logs even on success runs.

**Comment by as-com** [comment]
@dcbw done

**Comment by as-com** [comment]
Some tests are now failing after this rebase when they used to not fail. Some investigation may be necessary.

**Comment by trozet** [comment]
@as-com can you please rebase this? Underlying issue should be fixed now.

**Comment by as-com** [comment]
done @trozet

**Comment by trozet** [comment]
@dcbw PTAL

File: .github/workflow-templates/test.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ jobs:
         mkdir -p $GOPATH/src/github.com/ovn-org
         mv $GITHUB_WORKSPACE $GOPATH/src/github.com/ovn-org/ovn-kubernetes
         echo "::set-env name=WORKDIR::$GOPATH/src/github.com/ovn-org/ovn-kubernetes"
+        mkdir -p $GITHUB_WORKSPACE
```
Comment (girishmg) [inline .github/workflow-templates/test.yml]: @as-com @trozet at L32 we are doing a `mv`, but at L34 we are trying to create this directory? Am I missing something?

Comment (as-com) [reply .github/workflow-templates/test.yml]: We recreate this directory because some steps assume $GITHUB_WORKSPACE always exists.

**Pull request merged**